File: gcj_2015/QR_B.py
#!/usr/bin/env python

# Google Code Jam 2015 Qualification Round
# Problem B. Infinite House of Pancakes

import sys
import math

def solve(D, P):
  ans = max(P)
  for d in range(1, 1001):
    tot = d
    for p in P:
      if p > d:
        tot += int(math.ceil(float(p) / d)) - 1
    ans = min(ans, tot)
  return ans

T = int(sys.stdin.readline())
for t in range(1, T+1):
  D = int(sys.stdin.readline())
  P = [int(x) for x in sys.stdin.readline().rstrip("\n").split(' ')]
  ans = solve(D, P)
  print("Case #" + str(t) + ": " + str(ans))

File: gcj_2021/QR_B.cpp
// Google Code Jam 2021 Qualification Round
// Problem B. Moons and Umbrellas (5pts, 11pts, 1pts)

#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>

using namespace std;

int solve(int X, int Y, string s) {
	int ans = 0;
	vector<int> dp({ s[0] == 'J' ? 1 << 30 : 0, s[0] == 'C' ? 1 << 30 : 0 });
	for (int i = 1; i < (int)s.length(); ++i) {
		vector<int> next(2, 1 << 30);
		if (s[i] == 'C' || s[i] == '?') {
			next[0] = min(dp[0], dp[1] + Y);
		}
		if (s[i] == 'J' || s[i] == '?') {
			next[1] = min(dp[1], dp[0] + X);
		}
		dp = next;
	}
	return min(dp[0], dp[1]);
}

int main(int argc, char *argv[]) {
	int T, X, Y;
	cin >> T;
	for (int t = 1; t <= T; ++t) {
		string s;
		cin >> X >> Y >> s;
		int ans = solve(X, Y, s);
		cout << "Case #" << t << ": " << ans << endl;
	}
	return 0;
}
